File: ruby/bowling.rb
class Game

  def initialize
    @rolls = []
  end

  def roll(n)
    @rolls << n
  end

  def get_score()
    if @score != 0
      @score = 0
    end

    mid_frame = false
    is_strike = false

    @rolls.each_index do |i|
      roll = @rolls[i]

      if roll == 10 or mid_frame
        if is_strike
          if mid_frame
            @score += roll + @rolls[i-1]
          else
            if i < @rolls.count - 1
              @score += roll + @rolls[i+1]
            end
          end
          is_strike = false
        end

        if roll == 10 and not mid_frame
            @score += roll
            is_strike = true
        else
          if roll + @rolls[i-1] == 10 and mid_frame and i < @rolls.count - 1
            @score += @rolls[i+1]
          end

          @score += roll + @rolls[i-1]
        end
        mid_frame = false
      else
        mid_frame = true
      end

    end

    return @score
  end
end

File: javascript/bowling.js
var Game = function() {
  	this.score = 0;
    this.rolls = [];

    this.roll = function(n) {
        this.rolls.push(n);
    }

    this.getScore = function() {
        if(this.score !== 0) {
            this.score = 0;
        }

        var midFrame = false;
        var isStrike = false;

        for(i = 0; i < this.rolls.length; i++) {
            if(this.rolls[i] === 10 || midFrame) {
                if(isStrike) {
                    if(midFrame) {
                            this.score += this.rolls[i] + this.rolls[i - 1];
                    } else {
                        if(i != this.rolls.length - 1) {
                            this.score += this.rolls[i] + this.rolls[i + 1];
                        }
                    }
                    isStrike = false;
                }

                if(this.rolls[i] === 10 && !midFrame) {
                    this.score += this.rolls[i];
                    isStrike = true;
                } else {
                    if((this.rolls[i-1] + this.rolls[i] === 10) && midFrame){
                        if(i != this.rolls.length - 1) {
                            this.score += this.rolls[i + 1];
                        }
                    }
                    this.score += this.rolls[i] + this.rolls[i-1];
                    if (midFrame) {
                        midFrame = false;
                    }
                }
            } else {
                    midFrame = true;
            }
        }

        return this.score;
    }
}

module.exports = Game;